// _@antv_f2@3.1.12@@antv/f2/lib/interaction/base.js
function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

/**
 * Interaction 基类
 * @author sima.zhang1990@gmail.com
 */
var Util = require('../util/common');
var Chart = require('../chart/chart');

var Hammer = require('hammerjs');
Hammer = typeof Hammer === 'function' ? Hammer : window.Hammer;

var TOUCH_EVENTS = ['touchstart', 'touchmove', 'touchend'];

var Interaction = function () {
  Interaction.prototype.getDefaultCfg = function getDefaultCfg() {
    return {
      startEvent: TOUCH_EVENTS[0],
      processingEvent: TOUCH_EVENTS[1],
      endEvent: TOUCH_EVENTS[2],
      resetEvent: null
    };
  };

  Interaction.prototype._start = function _start(ev) {
    this.preStart && this.preStart(ev);
    this.start(ev);
    this.onStart && this.onStart(ev);
  };

  Interaction.prototype._process = function _process(ev) {
    this.preProcess && this.preProcess(ev);
    this.process(ev);
    this.onProcess && this.onProcess(ev);
  };

  Interaction.prototype._end = function _end(ev) {
    this.preEnd && this.preEnd(ev);
    this.end(ev);
    this.onEnd && this.onEnd(ev);
  };

  Interaction.prototype._reset = function _reset(ev) {
    this.preReset && this.preReset(ev);
    this.reset(ev);
    this.onReset && this.onReset(ev);
  };

  // override


  Interaction.prototype.start = function start() {};
  // override


  Interaction.prototype.process = function process() {};
  // override


  Interaction.prototype.end = function end() {};
  // override


  Interaction.prototype.reset = function reset() {};

  function Interaction(cfg, chart) {
    _classCallCheck(this, Interaction);

    var defaultCfg = this.getDefaultCfg();
    Util.deepMix(this, defaultCfg, cfg);
    this.chart = chart;
    this.canvas = chart.get('canvas');
    this.el = chart.get('canvas').get('el');
    this._bindEvents();
  }

  Interaction.prototype._bindEvents = function _bindEvents() {
    this._clearEvents(); // clear events
    var startEvent = this.startEvent,
        processingEvent = this.processingEvent,
        endEvent = this.endEvent,
        resetEvent = this.resetEvent,
        el = this.el;

    var hammer = new Hammer(el);
    this.hammer = hammer;
    this._bindEvent(startEvent, '_start', hammer);
    this._bindEvent(processingEvent, '_process', hammer);
    this._bindEvent(endEvent, '_end', hammer);
    this._bindEvent(resetEvent, '_reset', hammer);
  };

  Interaction.prototype._clearEvents = function _clearEvents() {
    var hammer = this.hammer;
    var startEvent = this.startEvent,
        processingEvent = this.processingEvent,
        endEvent = this.endEvent,
        resetEvent = this.resetEvent;


    if (hammer) {
      hammer.destroy();
      this._clearTouchEvent(startEvent, '_start');
      this._clearTouchEvent(processingEvent, '_process');
      this._clearTouchEvent(endEvent, '_end');
      this._clearTouchEvent(resetEvent, '_reset');
      this.hammer = null;
    }
  };

  Interaction.prototype._bindEvent = function _bindEvent(eventName, methodName, hammer) {
    var el = this.el;
    if (eventName) {
      if (TOUCH_EVENTS.indexOf(eventName) !== -1) {
        Util.addEventListener(el, eventName, Util.wrapBehavior(this, methodName));
      } else {
        hammer.on(eventName, Util.wrapBehavior(this, methodName));
      }
    }
  };

  Interaction.prototype._clearTouchEvent = function _clearTouchEvent(eventName, methodName) {
    var el = this.el;
    if (eventName && TOUCH_EVENTS.indexOf(eventName) !== -1) {
      Util.removeEventListener(el, eventName, Util.getWrapBehavior(this, methodName));
    }
  };

  Interaction.prototype.destroy = function destroy() {
    this._clearEvents();
  };

  return Interaction;
}();

Chart._Interactions = {};
Chart.registerInteraction = function (type, constructor) {
  Chart._Interactions[type] = constructor;
};
Chart.getInteraction = function (type) {
  return Chart._Interactions[type];
};

Chart.prototype.interaction = function (type, cfg) {
  var interactions = this._interactions || {};
  if (interactions[type]) {
    // 如果重复注册，会将上一个交互实例销毁删除
    interactions[type].destroy();
  }
  var Ctor = Chart.getInteraction(type);
  var interact = new Ctor(cfg, this);
  interactions[type] = interact;
  this._interactions = interactions;
  return this;
};
Chart.prototype.clearInteraction = function (type) {
  var interactions = this._interactions;
  if (!interactions) return;
  if (type) {
    interactions[type] && interactions[type].destroy();
    delete interactions[type];
  } else {
    Util.each(interactions, function (interaction, key) {
      interaction.destroy();
      delete interactions[key];
    });
  }

  return this;
};

module.exports = Interaction;
