// _@antv_f2@3.1.12@@antv/f2/lib/interaction/interval-select.js
function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var Util = require('../util/common');
var Helper = require('../util/helper');
var Interaction = require('./base');
var Chart = require('../chart/chart');

var IntervalSelect = function (_Interaction) {
  _inherits(IntervalSelect, _Interaction);

  function IntervalSelect() {
    _classCallCheck(this, IntervalSelect);

    return _possibleConstructorReturn(this, _Interaction.apply(this, arguments));
  }

  IntervalSelect.prototype.getDefaultCfg = function getDefaultCfg() {
    var defaultCfg = _Interaction.prototype.getDefaultCfg.call(this);
    return Util.mix({}, defaultCfg, {
      startEvent: 'tap',
      processingEvent: null,
      selectStyle: {
        fillOpacity: 1
      }, // 被选中图形的样式
      unSelectStyle: {
        fillOpacity: 0.4
      }, // 未被选中图形的样式
      cancelable: true // 选中之后是否允许取消选中，默认允许取消选中
    });
  };

  IntervalSelect.prototype._reset = function _reset() {
    if (!this.selectedShape) {
      return;
    }
    var chart = this.chart;
    var geom = chart.get('geoms')[0];
    var container = geom.get('container');
    var children = container.get('children');

    Util.each(children, function (child) {
      var originAttrs = child.get('_originAttrs');
      if (originAttrs) {
        child._attrs.attrs = originAttrs;
        child.set('_originAttrs', null);
      }
      child.set('_selected', false);
    });
    this.canvas.draw();
  };

  IntervalSelect.prototype.start = function start(ev) {
    var chart = this.chart;
    if (ev.type === 'tap') {
      ev.clientX = ev.center.x;
      ev.clientY = ev.center.y;
    }

    var _Util$createEvent = Util.createEvent(ev, chart),
        x = _Util$createEvent.x,
        y = _Util$createEvent.y;

    var plot = chart.get('plotRange');
    if (!Helper.isPointInPlot({ x: x, y: y }, plot)) {
      // 不在绘图区域
      this._reset();
      return;
    }

    // 查找被点击的 shapw
    var geom = chart.get('geoms')[0];
    var container = geom.get('container');
    var children = container.get('children');
    var selectedShape = void 0;
    var unSelectedShapes = [];

    Util.each(children, function (child) {
      var box = child.getBBox();
      if (x >= box.x && x <= box.x + box.width && y >= box.y && y <= box.height + box.y) {
        // inbox
        selectedShape = child;
      } else {
        unSelectedShapes.push(child);
      }
    });

    if (selectedShape) {
      // 有图形被选中
      this.selectedShape = selectedShape;
      if (selectedShape.get('_selected')) {
        // 已经被选中
        if (!this.cancelable) {
          // 不允许取消选中则不处理
          return;
        }
        this._reset(); // 允许取消选中
      } else {
        // 未被选中
        var selectStyle = this.selectStyle,
            unSelectStyle = this.unSelectStyle;


        if (!selectedShape.get('_originAttrs')) {
          var originAttrs = Object.assign({}, selectedShape.attr());
          selectedShape.set('_originAttrs', originAttrs);
        }

        selectedShape.attr(selectStyle);

        Util.each(unSelectedShapes, function (child) {
          if (!child.get('_originAttrs')) {
            var _originAttrs = Object.assign({}, child.attr());
            child.set('_originAttrs', _originAttrs);
          } else {
            child.attr(child.get('_originAttrs'));
          }
          child.set('_selected', false);
          unSelectStyle && child.attr(unSelectStyle);
        });

        selectedShape.set('_selected', true);
        this.canvas.draw();
      }
    } else {
      // 没有选中图形，恢复原态
      this._reset();
      this.selectedShape = null;
    }
  };

  IntervalSelect.prototype.end = function end(ev) {
    var selectedShape = this.selectedShape;
    if (selectedShape && !selectedShape.get('destroyed')) {
      ev.data = selectedShape.get('origin')._origin; // 绘制数据，包含原始数据啊
      ev.shapeInfo = selectedShape.get('origin');
      ev.shape = selectedShape;
      ev.selected = !!selectedShape.get('_selected'); // 返回选中的状态
    }
  };

  IntervalSelect.prototype.reset = function reset() {
    this._reset();
  };

  return IntervalSelect;
}(Interaction);

Chart.registerInteraction('interval-select', IntervalSelect);
module.exports = IntervalSelect;
